src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormat.cs

```diff
@@ -738,16 +723,54 @@ private static bool IsUseGenitiveForm(ReadOnlySpan<char> format, int index, int
                         // character rule.
                         // That is, if we ask everyone to use single quote or double quote to insert characters,
                         // then we can remove this default block.
-                        result.Append(ch);
+                        result.Append(TChar.CreateTruncating(ch));
                         tokenLen = 1;
                         break;
                 }
                 i += tokenLen;
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void AppendChar<TChar>(ref ValueListBuilder<TChar> result, char ch) where TChar : unmanaged, IBinaryInteger<TChar>
+        {
+            if (typeof(TChar) == typeof(char) || char.IsAscii(ch))
+            {
+                result.Append(TChar.CreateTruncating(ch));
+            }
+            else
+            {
+                Debug.Assert(typeof(TChar) == typeof(byte));
+                var r = new Rune(ch);
+                r.EncodeToUtf8(MemoryMarshal.AsBytes(result.AppendSpan(r.Utf8SequenceLength)));
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void AppendString<TChar>(ref ValueListBuilder<TChar> result, scoped ReadOnlySpan<char> s) where TChar : unmanaged, IBinaryInteger<TChar>
+        {
+            if (typeof(TChar) == typeof(char))
+            {
+                result.Append(MemoryMarshal.Cast<char, TChar>(s));
+            }
+            else
+            {
+                Debug.Assert(typeof(TChar) == typeof(byte));
+                Encoding.UTF8.GetBytes(s, MemoryMarshal.Cast<TChar, byte>(result.AppendSpan(Encoding.UTF8.GetByteCount(s))));
+            }
+        }
+
+        internal static void FormatFraction<TChar>(ref ValueListBuilder<TChar> result, int fraction, ReadOnlySpan<char> fractionFormat) where TChar : unmanaged, IBinaryInteger<TChar>
+        {
+            // TODO: Update when Int32 implements IUtf8SpanFormattable
```

Comment:
Could we ensure a corresponding issue is logged for this? Very easy to miss/forget about `TODO` comments otherwise.

Reply:
Or just throw/assert if we get wrong type.